### PR TITLE
fix pipe UV_HANDLE_EMULATE_IOCP implementation

### PR DESCRIPTION
**Pull request by vtjnash**
UV_HANDLE_EMULATE_IOCP was missing a bit of code, and any tests.

Fixes: #629
CI: https://ci.nodejs.org/job/libuv-test-commit/1709/

### CODE REVIEW

**Comment by vtjnash** [comment]
PTAL! I know it looks like a lot of code movement, but if you look at each commit separately, I think each should look pretty straightforward.

**Comment by vtjnash** [comment]
Review bump!

File: src/win/pipe.c [hunk 1]
```diff
@@ -326,6 +327,36 @@ static int uv_set_pipe_handle(uv_loop_t* loop,
 }
 
 
+static int pipe_alloc_accept(uv_loop_t* loop, uv_pipe_t* handle,
```
Comment (saghul) [inline src/win/pipe.c]: style: can you line these up?

File: src/win/pipe.c [hunk 2]
```diff
@@ -326,6 +327,36 @@ static int uv_set_pipe_handle(uv_loop_t* loop,
...
+  uv_pipe_accept_t* req, BOOL firstInstance) {
+  assert(req->pipeHandle == INVALID_HANDLE_VALUE);
+
+  req->pipeHandle = CreateNamedPipeW(handle->name,
```
Comment (saghul) [inline src/win/pipe.c]: style: can you align the arguments to CreateNamedPipeW?

File: src/win/pipe.c [hunk 3]
```diff
@@ -326,6 +327,36 @@ static int uv_set_pipe_handle(uv_loop_t* loop,
...
+      PIPE_ACCESS_DUPLEX | FILE_FLAG_OVERLAPPED | WRITE_DAC |
+      (firstInstance ? FILE_FLAG_FIRST_PIPE_INSTANCE : 0),
+      PIPE_TYPE_BYTE | PIPE_READMODE_BYTE | PIPE_WAIT,
+      PIPE_UNLIMITED_INSTANCES, 65536, 65536, 0, NULL);
+
+  if (req->pipeHandle == INVALID_HANDLE_VALUE)
```
Comment (saghul) [inline src/win/pipe.c]: style: add braces

File: src/win/pipe.c [hunk 4]
```diff
@@ -326,6 +327,36 @@ static int uv_set_pipe_handle(uv_loop_t* loop,
...
+
+  /* Associate it with IOCP so we can get events. */
+  if (CreateIoCompletionPort(req->pipeHandle,
+                             loop->iocp,
+                             (ULONG_PTR) handle,
+                             0) == NULL) {
+    uv_fatal_error(GetLastError(), "CreateIoCompletionPort");
+  }
+
+  /* Stash a handle in the server object for use from places such as
```
Comment (saghul) [inline src/win/pipe.c]: IS this `handle` here the server objecct? If so, won't this be overriden when multiple connections happen?

File: src/win/pipe.c [hunk 5]
```diff
@@ -577,11 +596,6 @@ int uv_pipe_bind(uv_pipe_t* handle, const char* name) {
     handle->name = NULL;
   }
 
-  if (handle->pipe.serv.accept_reqs[0].pipeHandle != INVALID_HANDLE_VALUE) {
-    CloseHandle(handle->pipe.serv.accept_reqs[0].pipeHandle);
```
Comment (saghul) [inline src/win/pipe.c]: When is this handle closed now?

File: src/win/pipe.c [hunk 6]
```diff
@@ -827,29 +841,11 @@ static void uv_pipe_queue_accept(uv_loop_t* loop, uv_pipe_t* handle,
...
-
-    if (req->pipeHandle == INVALID_HANDLE_VALUE) {
-      SET_REQ_ERROR(req, GetLastError());
-      uv_insert_pending_req(loop, (uv_req_t*) req);
-      handle->reqs_pending++;
-      return;
-    }
-
-    if (uv_set_pipe_handle(loop, handle, req->pipeHandle, -1, 0)) {
-      CloseHandle(req->pipeHandle);
```
Comment (saghul) [inline src/win/pipe.c]: Why are we not closing the handle immediately now?

File: src/win/pipe.c [hunk 7]
```diff
@@ -925,6 +921,7 @@ int uv_pipe_accept(uv_pipe_t* server, uv_stream_t* client) {
     req->next_pending = NULL;
     req->pipeHandle = INVALID_HANDLE_VALUE;
 
+    server->handle = INVALID_HANDLE_VALUE;
```
Comment (saghul) [inline src/win/pipe.c]: Shouldn't server->handle be the actual server'ss pipe handle? Why are we null-ing it now?

File: src/win/pipe.c [hunk 8]
```diff
@@ -1381,11 +1390,6 @@ static int uv__pipe_write_data(uv_loop_t* loop,
     handle->write_queue_size += req->u.io.queued_bytes;
   } else if (handle->flags & UV_HANDLE_BLOCKING_WRITES) {
     /* Using overlapped IO, but wait for completion before returning */
-    req->u.io.overlapped.hEvent = CreateEvent(NULL, 1, 0, NULL);
```
Comment (saghul) [inline src/win/pipe.c]: Looks like we were creating an event with the manual reset flag sset before, but I see no call to ResetEvent, so was this a bug then?

Comment (vtjnash) [reply src/win/pipe.c]: No, you're looking at the wrong commit. This is just supposed to consolidate the repeated code. I see what you mean though about the parameter change there, we are reseting this by calling `CloseHandle`, but it does bring up some other bugs though, so I'll address those.

File: src/win/pipe.c (hunk 4 above)
Comment (vtjnash) [reply src/win/pipe.c]: Kinda: this `handle` is a placeholder for the the server object currently, and later will become the client object when we get a connection, at which point we need to allocate a new server object (and put that new handle here).

File: src/win/pipe.c (hunk 5 above)
Comment (vtjnash) [reply src/win/pipe.c]: There's no longer any codepaths that could reach here (actually, there wasn't before either, but that was obscured by the fact that `uv_set_pipe_handle` could return error codes from codepaths unreachable from this particular call)

File: src/win/pipe.c (hunk 6 above)
Comment (vtjnash) [reply src/win/pipe.c]: This code should have been unreachable (the errors were checking that the flags to CreateNamedPipeW were correct—but we just set those), so it was just a convoluted wrapper for calling `CreateIoCompletionPort`. We can now do just that more simply and directly (by defining the `pipe_alloc_accept` helper).

File: src/win/pipe.c (hunk 7 above)
Comment (vtjnash) [reply src/win/pipe.c]: The server doesn't have a permanent handle of its own, so it just temporarily borrows the handle for a future client. We just gave away our handle (`req = server->pipe.serv.pending_accepts; pipe_client->handle = req->pipeHandle;`), so we want to also drop our reference to it at the same time for clarity (`uv_pipe_queue_accept` will allocate a new one, if we're not closing).

**Comment by vtjnash** [comment]
OK, new changes are in separate commits (some style fixes, some fixup to squash on merge). new CI: https://ci.nodejs.org/job/libuv-test-commit/1736/

File: src/win/pipe.c (hunk 8 above)
Comment (saghul) [reply src/win/pipe.c]: Gotcha, thanks for explaining!

**Comment by saghul** [comment]
@vtjnash Want me to land this one for you?

**Comment by vtjnash** [comment]
That'd be great, I haven't had time recently to be able to devote to libuv.

**Comment by saghul** [comment]
Squashed the fixups and landed. Cheers @vtjnash !

**Pull request closed without merging**